### PR TITLE
[Serve] Use Async Handle for DAG Execution

### PR DESCRIPTION
**Pull request by simon-mo**
Signed-off-by: simon-mo <simon.mo@hey.com>

## Why are these changes needed?
This PR replaces the sync handle usage in deployment graph with async handle. It uses the following method
- Change async handle's `.remote` return value as a `asyncio.Task` instead of a `coroutine`. This doesn't change the UX `ref = await handle.remote()` but it allows the result to be awaited twice by DAG code.
- DAG still synchronously call `.execute` and get the `asyncio.Task` back. 
- The tasks are propagated directly to downstream's `.remote` call. 
- In the router's execution path, we use `PyObjScanner` to find all occurrence of `asyncio.Task` and resolve them in batch.
- Generalized `PyObjScanner` to handle arbitrary source type for replacement.
- Improved performance of `PyObjScanner` so it will not serialize input objects.

This PR is code complete and ready for review. Before merge, I'll update the documentation and example code. Add more tests. 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
cc @iycheng @jiaodong @suquark for PyObjScanner changes.

File: python/ray/dag/py_obj_scanner.py [hunk 1]
```diff
@@ -30,50 +34,78 @@ def _get_node(instance_id: int, node_index: int) -> DAGNodeBase:
...
+        if obj is _get_node or obj is _get_object:
+            # Only fall back to cloudpickle for these two functions.
+            return super().reducer_override(obj)
+        elif isinstance(obj, self.source_type):
             index = len(self._found)
             self._found.append(obj)
             return _get_node, (id(self), index)
+        else:
+            index = len(self._objects)
+            self._objects.append(obj)
```
Comment (sihanwang41) [inline python/ray/dag/py_obj_scanner.py:88]: how do we do the dict object to do append?

**Review by edoakes — APPROVED** [review]
LGTM. Thank you for adding unit tests to the PyObjScanner, that class makes me a bit nervous in general.

Only concern is exposing this `RayServeLazyAsyncHandle` as a user-facing class... if we can fix that for the release, great, but not a complete dealbreaker.

**Comment by simon-mo** [comment]
microbenchmark seems to be normal range. 
[before](https://buildkite.com/ray-project/release-tests-branch/builds/858#018265cc-db41-4fd3-97fc-d0112cc585e5/654-684) 
[after](https://buildkite.com/ray-project/release-tests-pr/builds/12592#01826a08-426a-42be-9d14-1bccf2453f04/694-724)
percent change:
```
0.13
-2.24
-1.68
0.52
2.43
1.61
1.24
0.36
2.51
-4.45
-4.46
-0.00
3.67
0.93
3.99
-3.16
1.95
-4.82
-0.45
-1.26
2.91
-1.53
0.60
-4.42
```

**Pull request merged**